### PR TITLE
Add more precommits

### PR DESCRIPTION
**Pull request by arjxn-py**
# Description

Suggested in [#3180(comment)](https://github.com/pybamm-team/PyBaMM/pull/3180#issuecomment-1646698735) & [#3180(comment)](https://github.com/pybamm-team/PyBaMM/pull/3180#issuecomment-1648059280), this PR aims to add some more default pre-commit hooks.

Fixes # (issue)

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all`
- [ ] The documentation builds: `$ python run-tests.py --doctest`

You can run unit and doctests together at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Review by Saransh-cpp — COMMENTED** [review]
@arjxn-py your branch is behind develop; hence, the failure. You'll have to merge develop. Also, feel free to add other pygrep hooks and disallow-caps in this PR itself.

**Comment by arjxn-py** [comment]
Ahh thanks @Saransh-cpp I was wondering why it wasn't giving conflicts on PR itself but in pre-commit CI, cool it was just a sync thing 🙂

**Comment by arjxn-py** [comment]
Now I wonder it it require some special permission here in CI to fix files inside `.github/workflow/...` because It is fixing them locally

**Comment by Saransh-cpp** [comment]
> GitHub prevented pre-commit.ci from autofixing this pr due to autofixes to a workflow file

You'll have to run pre-commit locally 🙂

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -21,3 +21,24 @@ repos:
...
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.4.0
+    hooks:
+        - id: check-added-large-files
+        - id: check-case-conflict
+        - id: check-merge-conflict
+        - id: check-yaml
+        - id: debug-statements
+        - id: end-of-file-fixer
+          exclude: ^docs ^.github/workflows
```
Comment (Saransh-cpp) [inline .pre-commit-config.yaml:34]: Why exclude workflows and docs?

File: README.md [hunk 2]
```diff
@@ -11,6 +11,7 @@
 [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/pybamm-team/PyBaMM/blob/main/)
 [![DOI](https://zenodo.org/badge/DOI/10.5334/jors.309.svg)](https://doi.org/10.5334/jors.309)
 [![release](https://img.shields.io/github/v/release/pybamm-team/PyBaMM?color=yellow)](https://github.com/pybamm-team/PyBaMM/releases)
+[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit)](https://github.com/pre-commit/pre-commit)
```
Comment (Saransh-cpp) [inline README.md:14]: This is maybe not required 🙂

**Review by Saransh-cpp — CHANGES_REQUESTED** [review]
Thanks, @arjxn-py! You'll have to rewrite the git history. I see that there are multiple commits with pre-commit changes, but there should be only 1, with its hash written down in the git-revs file.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (arjxn-py) [reply .pre-commit-config.yaml]: Not excluding now.

**Review by Saransh-cpp — COMMENTED** [review]
Thanks, @arjxn-py! Looks great! You'll still need to figure out how to fix the git history, for example, there are `pre-commit` changes in https://github.com/pybamm-team/PyBaMM/pull/3192/commits/13964f91b366165d21ef529e40a13159f1da0e6c and https://github.com/pybamm-team/PyBaMM/pull/3192/commits/96fc3872c8dc2688e24705e87caccea5fe7abd64, and they should not be there in these commits.

**Review by Saransh-cpp — APPROVED** [review]

**Pull request merged**